Issue #600: Improve DATEXXX

hawkfish opened the pull request:
Clean up the code and add special cases for the common situation where the part is a constant string.

Commit:
Issue #600: Improve DATEPART

Now that there  are structs for all temporal types,
we can dispense with the separate structs and lose
some duplicated code.

Commit:
Issue #600: Improve DATEPART

Special case constant date part strings.

Commit:
Issue #600: Improve DATETRUNC

Special case constant part string.
Refactor to look like the other part functions.

Commit:
Merge branch 'master' into hawkfish-datepart

Mytherin commented on src/function/scalar/date/date_trunc.cpp:
```diff
@@ -11,216 +11,304 @@
+		template <class TA, class TR>
+		static inline TR Operation(TA input) {
+			date_t date = Timestamp::GetDate(input);
+			return Timestamp::FromDatetime(Date::FromDate(Date::ExtractYear(date), Date::ExtractMonth(date), 1),
```
Using `Date::Convert` here instead of ExtractYear + ExtractMonth should save some cycles :)

hawkfish replied:
Good call - I wasn't looking at that stuff.

Mytherin commented on src/function/scalar/date/date_part.cpp:
```diff
@@ -372,485 +372,397 @@ struct DateDatePart {
+
+	if (part_arg.GetVectorType() == VectorType::CONSTANT_VECTOR) {
+		// Common case of constant part.
+		if (ConstantVector::IsNull(part_arg)) {
```
For `date_part` this case is already handled through the `DatePartSimplificationRule` rule, which converts date_part with a constant into individual functions (e.g.  `EXTRACT('YEAR' FROM time)` becomes `YEAR(time)`). Perhaps not worth the added code size of having this case handled here separately as well?

hawkfish replied:
Oh right. There were other things worth keeping in here, though, including the general cleanup to match the other "part" function internals and adding support for `TIME` types. I've also updated the documentation to include all the missing functions and types.

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some comments:

Commit:
Issue #600: Improve DATEXXX

Remove constant optimisation in DATEPART because the translator handles it.
Tweak Quater extraction to use Date::Convert.

hawkfish commented:
I expect removing the constant optimisation should implode the coverage statistics...

Mytherin commented on src/function/scalar/date/date_trunc.cpp:
```diff
@@ -11,216 +11,303 @@
+static void DateTruncUnaryExecutor(DatePartSpecifier type, Vector &left, Vector &result, idx_t count) {
+	switch (type) {
+	case DatePartSpecifier::MILLENNIUM:
+		UnaryExecutor::Execute<TA, TR, DateTrunc::MillenniumOperator>(left, result, count);
```
`MILLENNIUM, CENTURY, DECADE, YEAR, DAY, HOUR, MILLISECONDS` appear to be uncovered in the constant case (the `date_trunc` test case only handles these cases with non-constant specifiers). Could you add some tests for those?

hawkfish replied:
Milliseconds was actually incorrect, and microseconds was missing.

Mytherin reviewed (commented):
Thanks, looks great now! Two more minor coverage requests, then I think it's ready to go:

hawkfish commented on src/function/scalar/date/date_part.cpp:
```diff
@@ -372,485 +372,327 @@ struct DateDatePart {
+
+template <>
+int64_t DatePart::YearWeekOperator::Operation(dtime_t input) {
+	throw NotImplementedException("\"time\" units \"yearweek\" not recognized");
```
`yearweek` is missing from the date part codes and only registers as as function.

hawkfish replied:
Well that was a saga!

Commit:
Issue #600: Improve DATEXXX

Add missing DATE_TRUNC constant tests.
Fix DATE_TRUNC ms/µs confusion.
Add missing YEARWEEK date part and tests.

Mytherin commented:
Thanks, looks great!
